Feature: Add SASL authentication support

victorlcm opened the pull request:
Hi 👋 ,

I've started using this extension as the built in kakfa output is being deprecated. I had a requirement to connect in a topic using SASL authentication, but It was not possible to do that using the current implementation of the kafka output. So, I've added the support in a fork and thought it would be a great thing to add in the extension, as other people might need this 😄

MStoykov commented on pkg/kafka/config.go:
```diff
@@ -78,6 +85,11 @@ func (c Config) Apply(cfg Config) Config {
 	if cfg.PushInterval.Valid {
 		c.PushInterval = cfg.PushInterval
 	}
+	if cfg.AuthMechanism != null.StringFrom("none") && cfg.User.Valid && cfg.Password.Valid {
+		c.AuthMechanism = cfg.AuthMechanism
+		c.User = cfg.User
+		c.Password = cfg.Password
+	}
```
IMO this needs to be 3 different ifs and than the config should be validated that if it has user, password or auth mechanism configured, all the others should be as well.

MStoykov replied:
for example in the `GetConsolidatedConfig`, maybe by adding a separate Validate method as well.

victorlcm replied:
Done

MStoykov commented on go.mod:
```diff
@@ -3,13 +3,15 @@ module github.com/k6io/xk6-output-kafka
 	github.com/mitchellh/mapstructure v1.1.2
 	github.com/sirupsen/logrus v1.8.1
 	github.com/stretchr/testify v1.7.0
+	github.com/xdg/scram v1.0.3
+	github.com/xdg/stringprep v1.0.3 // indirect
```
This dependency seems to have moved to `github.com/xdg-go` in 2018 but `sarama` hasn't updated the paths :facepalm: .  

I don't think you can or should do anything just pointing it out for other reviewers.

MStoykov commented on pkg/kafka/collector.go:
```diff
@@ -51,7 +52,35 @@ func New(p output.Params) (*Collector, error) {
+		}
+	}
+
+	saramaConfig.Producer.Return.Successes = true
```
Is this needed or used ?
From the [documentation](https://pkg.go.dev/github.com/Shopify/sarama#Config)

			// If enabled, successfully delivered messages will be returned on the
			// Successes channel (default disabled).
			Successes bool
			
This seems irrelevant for our needs ?

victorlcm replied:
As I've updated the sarama version, this is now required to be `true` for the SyncProducer. It's a little bit hidden in the SyncProducer [documentation](https://pkg.go.dev/github.com/Shopify/sarama#SyncProducer)

> For implementation reasons, the SyncProducer requires `Producer.Return.Errors` and `Producer.Return.Successes` to be set to true in its configuration.

MStoykov reviewed (changes requested):
Thanks for this PR :tada: and sorry for the late response.

I haven't tested this functionally as I don't really use kafka :grimacing: , but maybe @mostafa will like to help on that front.

I would really want you to rebase on the latest master(sorry we have been planning the changes that landed for months) and to have some of the configuration fixed.

Other than that it seems like a solid PR, good job :+1:

victorlcm commented on pkg/kafka/collector.go:
```diff
@@ -51,7 +52,35 @@ func New(p output.Params) (*Collector, error) {
+			saramaConfig.Net.SASL.SCRAMClientGeneratorFunc = func() sarama.SCRAMClient { return &XDGSCRAMClient{HashGeneratorFcn: SHA512} }
+		case "scram-sha-256":
+			saramaConfig.Net.SASL.Mechanism = sarama.SASLTypeSCRAMSHA256
+			saramaConfig.Net.SASL.SCRAMClientGeneratorFunc = func() sarama.SCRAMClient { return &XDGSCRAMClient{HashGeneratorFcn: SHA256} }
```
Done 😄

victorlcm commented:
Hey @MStoykov, I've implemented the requested changes 😄 
Thank you for the review!

MStoykov commented on pkg/kafka/config.go:
```diff
@@ -141,6 +158,11 @@ func GetConsolidatedConfig(jsonRawConf json.RawMessage, env map[string]string, a
 		return result, err
 	}
+
+	if envConfig.AuthMechanism.String != "none" && !envConfig.User.Valid && !envConfig.Password.Valid {
```
Shouldn't this be 
```suggestion
	if envConfig.AuthMechanism.String != "none" && (!envConfig.User.Valid || !envConfig.Password.Valid) {
```
? 

Although to be honest I do think the check should be that if one of them is set ... all the others need to be as well :thinking: but that will be a bit ... wordy 

Can you please also add some tests for this :grin:

victorlcm replied:
I've tested removing this whole condition to see what happened when one of the parameters was missing. Sarama already validates these parameters and shows more accurate errors, do you think we can stick with those?

e.g. 
```ERRO[0015] could not create the 'kafka' output: kafka: invalid configuration (Net.SASL.User must not be empty when SASL is enabled)```

victorlcm replied:
I've added some tests to cover this scenario 😄

victorlcm commented on go.mod:
```diff
@@ -3,20 +3,14 @@ module github.com/k6io/xk6-output-kafka
 	github.com/stretchr/testify v1.7.0
-	go.k6.io/k6 v0.33.1-0.20210701133132-aa1fd6a389c9
+	github.com/xdg/scram v1.0.3
+	go.k6.io/k6 v0.33.0
```
I see, I might need help fixing this. I tried using the specific commit, but it yielded an error when trying to run the local codebase using `xk6`

```2021/07/08 11:40:59 [INFO] exec (timeout=10s): /usr/local/go/bin/go mod init k6
go: creating new go.mod: module k6
go: to add module requirements and sums:
	go mod tidy
2021/07/08 11:40:59 [INFO] Replace github.com/k6io/xk6-output-kafka => /Users/victorlcm/Desktop/repos/xk6-output-kafka
2021/07/08 11:40:59 [INFO] exec (timeout=10s): /usr/local/go/bin/go mod edit -replace github.com/k6io/xk6-output-kafka=/Users/victorlcm/Desktop/repos/xk6-output-kafka
2021/07/08 11:40:59 [INFO] Pinning versions
2021/07/08 11:40:59 [INFO] exec (timeout=0s): /usr/local/go/bin/go get -d -v go.k6.io/k6
go get: go.k6.io/k6@upgrade (v0.33.0) requires go.k6.io/k6@v0.33.1-0.20210701133132-aa1fd6a389c9, not go.k6.io/k6@upgrade (v0.33.0)
2021/07/08 11:41:00 [ERROR] exit status 1
```

Do you know what can be causing this?

victorlcm replied:
hey @MStoykov did you have the chance to look into this issue? I'm having the same problem when trying to use xk6 to build or execute even from the master branch locally e.g. `xk6 build --with github.com/k6io/xk6-output-kafka=.` it yields the same error

MStoykov replied:
ooh, terribly sorry, this has apparently fallen through the cracks at some point and I don't remember seeing it at all :sob: 

At least for me 
`xk6 build --with github.com/k6io/xk6-output-kafka@master`

Works perfectly, but I also got the same error until I did 
```
K6_VERSION=v0.33.1-0.20210701133132-aa1fd6a389c9  xk6 build --with github.com/k6io/xk6-output-kafka=.
```
so it might be an issue with xk6 that I will look into.

Hope this helps you and sorry for very late reply

victorlcm replied:
No worries 😄  I was on holidays so I didn't follow up here for some days
your suggestion worked! I've updated the code with this specific version. Thank you very much!

mostafa commented:
Hi @victorlcm,

thanks for your contribution. I have chosen [kafka-go](https://github.com/segmentio/kafka-go) for my project, [xk6-kafka](https://github.com/mostafa/xk6-kafka), but the basics are the same. So, I suggest you have look at https://github.com/mostafa/xk6-kafka/pull/3 and specifically the [auth.go](https://github.com/mostafa/xk6-kafka/blob/master/auth.go) file. I'll definitely review it after you apply @MStoykov's suggestions.

victorlcm commented:
Hey @MStoykov , I've finished updating the code reviews and I've added some new extra tests. Do you think we're good to merge this one? Thanks!

MStoykov commented on pkg/kafka/collector.go:
```diff
@@ -69,8 +108,10 @@ func (c *Collector) Description() string {
 func (c *Collector) Stop() error {
-	c.done <- struct{}{}
-	<-c.done
+	go func() {
+		c.done <- struct{}{}
+		<-c.done
+	}()
```
I do think this needs to block so k6 waits for the output to flush everything

victorlcm replied:
Makes sense, I was having an issue because the job was staying alive for too long and this fixed it. But you're right, this is actually losing messages. Do you think it's good idea to make the push metrics run in go routines to speed up the flush process? 

```go
for {
  select {
  case <-ticker.C:
  go c.pushMetrics()
  case <-c.done:
  c.pushMetrics()
```

MStoykov replied:
if that is possible without leading to race conditions - yes, but not in this PR ;)

victorlcm replied:
sure, I'lll open another one =)

MStoykov commented on pkg/kafka/config.go:
```diff
@@ -38,22 +40,30 @@ type Config struct {
+	AuthMechanism null.String `json:"auth_mechanism" mapstructure:"auth_mechanism" envconfig:"K6_KAFKA_AUTH_MECHANISM"`
 
 	InfluxDBConfig influxdbConfig `json:"influxdb" mapstructure:"influxdb"`
+	Version        string         `json:"version" mapstructure:"version"`
```
what is the idea behind this Version? Isn't this something that maybe the kafka should return to us? Sorry I have near zero experience with kafka

victorlcm replied:
Sometimes it's necessary to specify the version to match the kafka deployment and get some updates. For instance, the minimum version does not send the timestamp automatically, which affects the ability to use ksql  to stream process the messages. In this case I was kinda forced to add the parameters due to a limitation in Sarama mocks. The mocks currently work only with the minimum version (0.8.2.0) and the default version in the library is 1.0.0.0, causing the tests to always fail 😄

MStoykov reviewed (approved):
I have no more comments on the code, but haven't tested it and am still somewhat skeptical of the need for `version` as it didn't need it so far ... or did it :thinking:

mostafa commented:
Well done @victorlcm!

@MStoykov I'll test it.

mostafa commented:
I tried to run a test and this is the result:

```bash
[appuser@broker /]$ ./k6 run -o xk6-kafka=brokers=localhost:9092.topic=k6,user=kafka,password=kafka,auth_mechanism=scram-sha-512,format=json script.js

          /\      |‾‾| /‾‾/   /‾‾/
     /\  /  \     |  |/  /   /  /
    /  \/    \    |     (   /   ‾‾\
   /          \   |  |\  \ |  (‾)  |
  / __________ \  |__| \__\ \_____/ .io

ERRO[0000] could not create the 'xk6-kafka' output: 3 error(s) decoding:

* 'auth_mechanism' expected a map, got 'string'
* 'password' expected a map, got 'string'
* 'user' expected a map, got 'string'
```

victorlcm commented:
Hey @mostafa,

Just fixed it, should be working now 😄

mostafa commented:
I am using kafka-docker-playground's [sasl env](https://github.com/vdesabou/kafka-docker-playground/tree/master/environment/sasl-scram) and this doesn't work:
```bash
./k6 run -o xk6-kafka=brokers=broker:9092,topic=k6,user=client,password=client-secret,auth_mechanism=scram-sha-512,format=json script.js

          /\      |‾‾| /‾‾/   /‾‾/
     /\  /  \     |  |/  /   /  /
    /  \/    \    |     (   /   ‾‾\
   /          \   |  |\  \ |  (‾)  |
  / __________ \  |__| \__\ \_____/ .io

ERRO[0001] could not create the 'xk6-kafka' output: kafka: client has run out of available brokers to talk to (Is your cluster reachable?)

Init      [--------------------------------------]
default   [--------------------------------------]
```

Kafka logs:
```
2021-08-20T08:30:17.671978365Z [2021-08-20 08:30:17,671] INFO [SocketServer listenerType=ZK_BROKER, nodeId=1] Failed authentication with /172.21.0.1 (errorMessage=Unexpected Kafka request of type METADATA during SASL handshake.) (org.apache.kafka.common.network.Selector)
```

victorlcm commented:
Hey @mostafa,

I've investigated the issue and apparently it's related to the security protocol. The env you're using is PLAINTEXT and not SSL. So, I've changed the default configs to use PLAINTEXT and added a SSL option. The Kafkas broker I was using had SSL enabled, that's why I was not getting this error. Thank you for pointing it out!

mostafa commented:
This worked and it successfully published messages to Kafka. Yay! 🎉

```bash
./k6 run -d 10s -o xk6-kafka=brokers=broker:9092,topic=k6,user=client,password=client-secret,auth_mechanism=scram-sha-256,format=json script.js

          /\      |‾‾| /‾‾/   /‾‾/
     /\  /  \     |  |/  /   /  /
    /  \/    \    |     (   /   ‾‾\
   /          \   |  |\  \ |  (‾)  |
  / __________ \  |__| \__\ \_____/ .io

  execution: local
     script: script.js
     output: kafka: TODO <<<<<< What is this?

  scenarios: (100.00%) 1 scenario, 1 max VUs, 40s max duration (incl. graceful stop):
           * default: 1 looping VUs for 10s (gracefulStop: 30s)


running (10.5s), 0/1 VUs, 9 complete and 0 interrupted iterations
default ✓ [======================================] 1 VUs  10s

     data_received..................: 108 kB 10 kB/s
     data_sent......................: 1.8 kB 171 B/s
     http_req_blocked...............: avg=67.31ms  min=371ns    med=842ns    max=605.84ms p(90)=121.17ms p(95)=363.5ms
     http_req_connecting............: avg=10.65ms  min=0s       med=0s       max=95.9ms   p(90)=19.18ms  p(95)=57.54ms
     http_req_duration..............: avg=99.96ms  min=98.23ms  med=100.16ms max=102.85ms p(90)=101.3ms  p(95)=102.07ms
       { expected_response:true }...: avg=99.96ms  min=98.23ms  med=100.16ms max=102.85ms p(90)=101.3ms  p(95)=102.07ms
     http_req_failed................: 0.00%  ✓ 0        ✗ 9
     http_req_receiving.............: avg=859.72µs min=298.59µs med=426.95µs max=2.73ms   p(90)=1.79ms   p(95)=2.26ms
     http_req_sending...............: avg=233.93µs min=159.89µs med=198.95µs max=493.25µs p(90)=293.31µs p(95)=393.28µs
     http_req_tls_handshaking.......: avg=27.27ms  min=0s       med=0s       max=245.49ms p(90)=49.09ms  p(95)=147.29ms
     http_req_waiting...............: avg=98.87ms  min=97.21ms  med=98.53ms  max=101.1ms  p(90)=100.13ms p(95)=100.62ms
     http_reqs......................: 9      0.855762/s
     iteration_duration.............: avg=1.16s    min=1.09s    med=1.1s     max=1.7s     p(90)=1.22s    p(95)=1.46s
     iterations.....................: 9      0.855762/s
     vus............................: 1      min=1      max=1
     vus_max........................: 1      min=1      max=1
```

mostafa commented on pkg/kafka/collector.go:
```diff
@@ -51,7 +53,48 @@ func New(p output.Params) (*Collector, error) {
+		if conf.SSL {
+			saramaConfig.Net.TLS.Enable = true
+			saramaConfig.Net.TLS.Config = &tls.Config{
+				InsecureSkipVerify: true,
+				ClientAuth:         0,
```
Probably not a good idea to skip certificate verification. It would be good if it's optional via a parameter, rather than being hardcoded.

victorlcm replied:
Hey @mostafa, I've just added a config, default is to not skip 😄

mostafa reviewed (changes requested):
Well done! Left just one comment.

victorlcm commented:
hey @MStoykov, can we merge this one? 😄

codebien commented on pkg/kafka/collector.go:
```diff
@@ -69,26 +112,28 @@ func (c *Collector) Description() string {
 }
 
 func (c *Collector) Stop() error {
+
```
```suggestion
```
it can be removed

codebien commented on pkg/kafka/scram_client.go:
```diff
@@ -0,0 +1,37 @@
+	SHA512 scram.HashGeneratorFcn = sha512.New
+)
+
+type XDGSCRAMClient struct {
```
this struct can be unexported

codebien reviewed (changes requested):
Hi @victorlcm,
thanks for your contribution. Added few comments.
